src/cmap/auth/gssapi.ts

```diff
@@ -21,65 +31,159 @@ export class GSSAPI extends AuthProvider {
+          if (err) return callback(err);
 
-            connection.command('$external.$cmd', command, (err, doc) => {
-              if (err) return callback(err, false);
+          externalCommand(saslContinue(payload, result.conversationId), (err, result) => {
+            if (err) return callback(err);
+            finalize(client, username, result.payload, (err, payload) => {
+              if (err) return callback(err);
 
-              authProcess.transition(doc.payload, (err, payload) => {
-                if (err) return callback(err, false);
-                const command = {
+              externalCommand(
+                {
                   saslContinue: 1,
-                  conversationId: doc.conversationId,
+                  conversationId: result.conversationId,
                   payload
-                };
+                },
+                (err, result) => {
+                  if (err) return callback(err);
 
-                connection.command('$external.$cmd', command, (err, response) => {
-                  if (err) return callback(err, false);
-
-                  authProcess.transition(null, err => {
-                    if (err) return callback(err);
-                    callback(undefined, response);
-                  });
-                });
-              });
+                  callback(undefined, result);
+                }
+              );
             });
           });
         });
       });
     });
   }
 }
+function saslStart(payload?: string): object {
+  return {
+    saslStart: 1,
+    mechanism: 'GSSAPI',
+    payload,
+    autoAuthorize: 1
+  };
+}
+
+function saslContinue(payload?: string, conversationId?: number): object {
+  return {
+    saslContinue: 1,
+    conversationId,
+    payload
+  };
+}
+
+function negotiate(
+  client: KerberosClient,
+  retries: number,
+  payload: string,
+  callback: Callback<string>
+) {
+  client.step(payload, (err, response) => {
+    // Retries exhausted, raise error
+    if (err && retries === 0) return callback(err);
+
+    // Adjust number of retries and call step again
+    if (err) return negotiate(client, retries - 1, payload, callback);
+
+    // Return the payload
+    callback(undefined, response || '');
+  });
+}
+
+function finalize(
+  client: KerberosClient,
+  user: string,
+  payload: string,
+  callback: Callback<string>
+) {
+  // GSS Client Unwrap
+  client.unwrap(payload, (err, response) => {
+    if (err) return callback(err);
+
+    // Wrap the response
+    client.wrap(response || '', { user }, (err, wrapped) => {
+      if (err) return callback(err);
+
+      // Return the payload
+      callback(undefined, wrapped);
+    });
+  });
+}
+
+function performGssapiCanonicalizeHostName(
+  canonicalizeHostName: boolean,
+  host: string,
+  callback: Callback<string>
+) {
+  if (!canonicalizeHostName) return callback(undefined, host);
+
+  // Attempt to resolve the host name
+  dns.resolveCname(host, (err: Error | null, r: string[]) => {
```

Comment:
can we drop the type annotations on the callback since we have types from the built-in module?